magma/circuit.py

```diff
@@ -524,28 +530,48 @@ def place(cls, inst):
             inst.stack = inspect.stack()
         cls.instances.append(inst)
 
+    def gen_bind_port(cls, mon_arg, bind_arg):
+        if isinstance(mon_arg, Tuple) or isinstance(mon_arg, Array) and \
+                not issubclass(mon_arg.T, m.Digital):
+            result = []
+            for child1, child2 in zip(mon_arg, bind_arg):
+                result += cls.gen_bind_port(child1, child2)
+            return result
+        else:
+            return [(f".{verilog_name(mon_arg.name)}({verilog_name(bind_arg.name)})")]
+
     def bind(cls, monitor, *args):
         bind_str = monitor.verilogFile
 
         ports = []
-        for mon_arg, cls_arg in zip(monitor.interface.ports,
-                                    cls.interface.ports):
-            if mon_arg != cls_arg:
+        for mon_arg, cls_arg in zip(monitor.interface.ports.values(),
+                                    cls.interface.ports.values()):
+            if str(mon_arg.name) != str(cls_arg.name):
                 error_str = f"""
 Bind monitor interface does not match circuit interface
     Monitor Ports: {list(monitor.interface.ports)}
     Circuit Ports: {list(cls.interface.ports)}
 """
                 raise TypeError(error_str)
-            ports.append(f".{mon_arg}({cls_arg})")
+            ports += cls.gen_bind_port(mon_arg, cls_arg)
         extra_mon_args = list(
-            monitor.interface.ports.keys()
+            monitor.interface.ports.values()
         )[len(cls.interface):]
         for mon_arg, bind_arg in zip(extra_mon_args, args):
-            ports.append(f".{mon_arg}({verilog_name(bind_arg.name)})")
+            ports += cls.gen_bind_port(mon_arg, bind_arg)
         ports_str = ", ".join(ports)
         bind_str = f"bind {cls.name} {monitor.name} {monitor.name}_inst ({ports_str});"  # noqa
-        cls.bind_modules.append((monitor, bind_str))
+        if not os.path.isdir(".magma"):
+            os.mkdir(".magma")
+        curr_compile_dir = get_compile_dir()
+        set_compile_dir("normal")
+        # Circular dependency, need coreir backend to compile, backend imports
+        # circuit (for wrap casts logic, we might be able to factor that out)
```

Comment:
One way around this circular dependency would be to run a pass to compile bind modules during the `compile` phase, but I think that's more work than is desired for a temporary solution that works with this circular import logic.

Reply:
You read my mind :). Can you add a note on the `import magma as m` that this is undesired and should be removed asap?